File: 백준/Gold/7579. 앱/앱.js
const readline = require('readline').createInterface({
  input: process.stdin,
  output: process.stdout
});

let input = [];
let N, M, memory, cost;

readline.on('line', (line) => {
  input.push(line);
}).on('close', () => {
  [N, M]= input[0].split(' ').map(Number);
  memory = input[1].split(' ').map(Number);
  cost = input[2].split(' ').map(Number);
  const result = solution();
  console.log(result);
  process.exit();
});

const solution = () => {
  let result = Infinity;
  const totalCost = cost.reduce((sum, c) => sum + c, 0);
  let DP = Array.from({ length: N }, () => new Array(totalCost+1).fill(0));

  for (let j = cost[0]; j <= totalCost; j++) {
    DP[0][j] = memory[0];
  }

  for (let i = 1; i < N; i++) {
    for (let j = 0; j <= totalCost; j++) {
      if (j-cost[i] >= 0) {
        DP[i][j] = Math.max(DP[i-1][j], DP[i-1][j-cost[i]]+memory[i]);
      } else {
        DP[i][j] = DP[i-1][j];
      }
    }
  }

  for (let i = 0; i <= totalCost; i++) {
    if (DP[N-1][i] >= M) {
      result = Math.min(result, i);
    }
  }
  
  return result;
}

File: 백준/Gold/7579. 앱/앱.py
import sys
input = sys.stdin.readline

n, m = map(int, input().split())
memory = list(map(int, input().split()))
cost = list(map(int, input().split()))

sum_of_cost = sum(cost)
dp = [[0 for _ in range(sum_of_cost+1)] for _ in range(n)] # dp[i][j] -> i번째 앱까지 봤을 때 j비용으로 확보한 메모리 크기

for j in range(cost[0], sum_of_cost+1):
  dp[0][j] = memory[0]

for i in range(1, n):
  for j in range(sum_of_cost+1):
    if j-cost[i] >= 0:
      dp[i][j] = max(dp[i-1][j], dp[i-1][j-cost[i]]+memory[i]) 
    else: 
      dp[i][j] = dp[i-1][j]

# n번째 앱에서 j비용으로 메모리를 확보한 크기를 확인하며, 만약 m과 같거나 넘는 순간이 오면 해당 비용을 return
for j in range(sum_of_cost+1):
  if dp[n-1][j] >= m:
    print(j)
    break